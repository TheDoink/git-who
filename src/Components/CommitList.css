#repoName {
	font-size:8vw;
	text-align: center;
	margin:15px;
}