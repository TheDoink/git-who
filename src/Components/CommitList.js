import React, { Component } from 'react';
import CommitItem from './CommitItem';
import './CommitList.css';

import $ from 'jquery';

class CommitList extends Component {

  constructor(props) {
    super(props);

    this.state = {
      commitList: [], // This list of commits for this repo
      pageNum: 1,     // The page number we're on (30 items per page)
      morePages: true,// If there are more pages to go (since we don't know the total number of commits)
      loading: false, // Whether or not we're in the process of getting more
      failure: false  // whether or not we've had a catastrophic page failure
    }
  };

  getCommits(gitUser, repoName, pageNum) {

    // Ensure we actually have a good user
    if(gitUser.login) {
      this.setState({loading: true}, function() {

        let since = new Date();
        since.setMonth(since.getMonth()-1);
        since = since.toISOString();

        // Try and get that username...
        $.get(`https://api.github.com/repos/${gitUser.login}/${repoName}/commits?since=${since}&page=${pageNum}`,

          // If we get a good response...
          (success, text, obj) => {

            let morePages = true;
            if(success.length < 30) {
              morePages = false;
            }

            this.state.commitList = this.state.commitList.concat(success);

            // Append the new repos to the old repos, and update the pageNum
            this.setState({commitList: this.state.commitList, pageNum: pageNum, failure: false, loading:false, morePages: morePages}, function() {
            });

          })

          // If we get anything else...
          .fail((failure, text, error) => {
            this.setState({failure: true, loading: false});
          });
      });
    }
  }

  /* Only get commits if we have a new username and a repo name*/
  componentWillReceiveProps(nextProps) {
    if(nextProps.gitUser && nextProps.repoName) {

      this.getCommits(nextProps.gitUser, nextProps.repoName, 1);
    }

    // Even if no good username, reset the list
    this.setState({commitList: [], pageNum: 1}, function() {});
  };

  render() {
    return (
      <div>
        {/* The repo title */}
        <div className="repoTitle">
          "{this.props.repoName}"
        </div>

        {/* Always Hide if there was a failure*/}
        <div className={this.state.failure}>

          {/* The list of commits */}
          {this.state.commitList.map(item => (
            <CommitItem key={item.sha} commit={item}>
              
            </CommitItem>
          ))}

          {/*Display is we have no items in our list*/}
          <div className={`noItems ${this.props.gitUser && !this.state.loading && this.state.commitList.length === 0 ? '' : 'hidden'}`}>
            <div>No commits to this repo in the last month!</div>
            <div><a href='/'>Back to {this.props.gitUser.login}</a></div>
          </div>

          {/* A button that is used to get additional repos, only shown when there is a valid user, and when not loading */}
          <div onClick={() => this.getCommits(this.props.gitUser, this.props.repoName, this.state.pageNum+1)} className={`getMore ${this.props.gitUser && !this.state.loading && this.state.morePages ? '' : 'hidden'}`}>
            <i className="fa fa-arrow-circle-o-down getMoreIcon"></i>
            <div className="getMoreText">Get More?</div>
          </div>
        </div>

        {/*Error message if there's a problem */}
        <div className={!this.state.failure ? 'hidden' : ''}>
          <div className="failText">
            A problem occured!
          </div>
        </div>
      </div>
    );
  }
}

export default CommitList;
