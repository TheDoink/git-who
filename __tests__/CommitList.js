// var NameLink = require('../src/Components/NameLink');
import React from 'react';
import renderer from 'react-test-renderer';
import CommitList from '../src/Components/CommitList';

const mockUser = {
      "login": "hank",
      "id": 8534,
      "avatar_url": "https://avatars1.githubusercontent.com/u/8534?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/hank",
      "html_url": "https://github.com/hank",
      "followers_url": "https://api.github.com/users/hank/followers",
      "following_url": "https://api.github.com/users/hank/following{/other_user}",
      "gists_url": "https://api.github.com/users/hank/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/hank/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/hank/subscriptions",
      "organizations_url": "https://api.github.com/users/hank/orgs",
      "repos_url": "https://api.github.com/users/hank/repos",
      "events_url": "https://api.github.com/users/hank/events{/privacy}",
      "received_events_url": "https://api.github.com/users/hank/received_events",
      "type": "User",
      "site_admin": false
    };

it('lines up with the snapshot', () => {
  let testMe = renderer.create(<CommitList repoName="git-who" gitUser={mockUser}></CommitList>).toJSON();

  expect(testMe).toMatchSnapshot();
  // testMe.getGist('d2eace7977e2aeaf942de0f8b3e7275c');

  
});
