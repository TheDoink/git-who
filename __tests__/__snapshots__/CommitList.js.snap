// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lines up with the snapshot 1`] = `
<div>
  <div
    className="repoTitle"
  >
    <span>
      <a
        href="#"
        onClick={[Function]}
      >
        hank
      </a>
    </span>
    /
    git-who
  </div>
  <div
    className={false}
  >
    <div
      className="noItems "
    >
      <div>
        No commits to this repo in the last month!
      </div>
      <div>
        <a
          href="/"
        >
          Back to 
          hank
        </a>
      </div>
    </div>
    <div
      className="getMore "
      onClick={[Function]}
    >
      <i
        className="fa fa-arrow-circle-o-down getMoreIcon"
      />
      <div
        className="getMoreText"
      >
        Get More?
      </div>
    </div>
  </div>
  <div
    className="hidden"
  >
    <div
      className="failText"
    >
      A problem occured!
    </div>
  </div>
</div>
`;
